### PR TITLE
Make space for tree sitter syntax highlighting

### PR DESCRIPTION
**Pull request by alexr00**
This PR takes the first steps towards adding tree sitter highlighting. I don't expect that there's much that's very interesting in here as there's no actual tree sitter based support in this PR, it's just about getting some of the uninteresting logistics out of the way to make reviewing subsequent PRs easier.

### CODE REVIEW

**Review by alexr00 — COMMENTED** [review]
@hediet if you would prefer to wait to merge or approve this until you see what the next part looks like, that's also fine for me.

File: src/vs/editor/common/model/tokenizationTextModelPart.ts [hunk 1]
```diff
@@ -44,7 +49,8 @@ export class TokenizationTextModelPart extends TextModelPart implements ITokeniz
 	private readonly _onDidChangeTokens: Emitter<IModelTokensChangedEvent> = this._register(new Emitter<IModelTokensChangedEvent>());
 	public readonly onDidChangeTokens: Event<IModelTokensChangedEvent> = this._onDidChangeTokens.event;
 
-	private readonly grammarTokens = this._register(new GrammarTokens(this._languageService.languageIdCodec, this._textModel, () => this._languageId, this._attachedViews));
+	private grammarTokens: GrammarTokens | undefined;
+	private treeSitterTokens: TreeSitterTokens | undefined;
```
Comment (hediet) [inline src/vs/editor/common/model/tokenizationTextModelPart.ts:53]: I'd suggest to not have a separate field `treeSitterTokens`, but have a tree sitter implemenation of `GrammarTokens` (and introducing an abstract class).

File: src/vs/workbench/services/treeSitter/browser/treeSitterTokenizationFeatureImpl.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+	constructor() {
+		// Eventually, this should actually use an extension point to add tree sitter grammars, but for now they are hard coded in core
+		this._addGrammar('placeholder-language');
+	}
+
+	private _addGrammar(languageId: string) {
+		TreeSitterTokenizationRegistry.register(languageId, {});
+	}
+}
```
Comment (hediet) [inline src/vs/workbench/services/treeSitter/browser/treeSitterTokenizationFeatureImpl.ts:20]: I think we should *really* try to make this available in the monaco-editor as well.
Both to make development easier (with the playground) and to invest into the monaco-editor (which doesn't even have official textmate support, which is very annoying).

**Review by hediet — CHANGES_REQUESTED** [review]
Looks good so far! Left some comments.
I love to iterate and merge quickly, but I think due to its explorative nature, we should make this a WIP PR and only merge it when the architecture is settled.

**Review by LakhveerChahal — APPROVED** [review]

File: src/vs/editor/common/model/tokenizationTextModelPart.ts (hunk 1 above)
Comment (alexr00) [reply src/vs/editor/common/model/tokenizationTextModelPart.ts:56]: Sounds good, I will address in an upcoming commit.

File: src/vs/editor/common/model/treeSitterTokens.ts [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+		private readonly _queries: string,
+		private readonly _language: Parser.Language,
+		private readonly _colorThemeData: ColorThemeData) {
+		super();
+		this._register(this._textModel.onDidChangeContent(() => this._onDidChangeContent()));
+	}
+
+	private _onDidChangeContent() {
+		// TODO @alexr00 this is an extremely naive implementation and should absolutely not be pushed to main, it's just a placeholder.
+		this._tree?.delete();
```
Comment (alexr00) [inline src/vs/editor/common/model/treeSitterTokens.ts:123]: @hediet this is just temporary. I will implement proper content change handling in an upcoming commit.

File: src/vs/editor/common/model/treeSitterTokens.ts [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+				}
+			}
+			tokens[tokenIndex * 2] = offset;
+			tokens[tokenIndex * 2 + 1] = metadata;
+			tokenIndex++;
+		}
+		return tokens;
+	}
+
+	private findMetadata(captureName: string): number {
```
Comment (alexr00) [inline src/vs/editor/common/model/treeSitterTokens.ts:187]: @hediet, do you know if there's a better way to retrieve the token metadata given a scope? Relying on the `ColorThemeData` is layer breaking. I started trying to move things around to fix the layer breaker, but it quickly got complicated.

**Pull request closed without merging**